### PR TITLE
implement open-ended reasoning

### PR DESCRIPTION
**Pull request by bddap**
This pr adds a new function. `infer`

`infer` doesn't have a goal in mind like `prove` does.
Unlike `validate`, `infer` thinks for itself. It generates its own conclusions.
`infer` is an open minded individual who won't stop until it groks its universe in full.

:)

### CODE REVIEW

File: src/common.rs [hunk 1]
```diff
@@ -19,6 +27,87 @@ where
     ret
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
```
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:30]: On it.

Unrelated observation, `lower` is kinda doing two separate things here. It both changes the way the rule is represented, and it translates elements from `T` to `usize`. Maybe `lower` simplified in the future by separating the two steps. The refactor would be a little tricky though.

File: src/common.rs [hunk 2]
```diff
@@ -19,6 +27,87 @@ where
...
+        }
+
+        RuleApplication {
+            rule_index: self.rule_index,
+            instantiations,
+        }
+    }
+}
+
+/// Attempt to forward translate a quad.
```
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:71]: Fixed in 55978e7

File: src/common.rs [hunk 3]
```diff
@@ -19,6 +27,87 @@ where
...
+            translator.forward(s)?,
+            translator.forward(p)?,
+            translator.forward(o)?,
+            translator.forward(g)?,
+        ]
+        .into(),
+    )
+}
+
+/// Reverse of forward.
```
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs]: Fixed in 55978e7

File: src/common.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:30]: Fixed in 55978e7

[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:30]: Still need to add docs for `LowRuleApplication`. Will do.

File: src/common.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:71]: Translator is a bi-directional dictionary from one language to another. One language is `T` and the other is `usize`.

```
forall t in T
  if trans.forward(t) == Some(u)
  then trans.back(u) == Some(t)
```

> btw does forward mean find the index of each of s,p,o, and g ?

If you want to call members of the second language (`usize`) indices then sure.
The fact that they are indices is an implementation detail. Translator currently looks like this:

```rust
pub struct Translator<T> {
    widdershins: Box<[T]>,
}
```

But a less efficient version of Translator might look like this:

```rust
pub struct Translator<T> {
    widdershins: BtreeMap<usize, T>,
    deosil: BtreeMap<T, usize>,
}
```

File: src/common.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs]: > and does this mean replace each index with the actual corresponding item ?

yes? I think? See above comment.

File: src/common.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bddap) [reply src/common.rs:30]: Done 26debb6

File: src/infer.rs [hunk 4]
```diff
@@ -184,4 +186,75 @@ mod tests {
         let expected: [[&str; 4]; 0] = [];
         assert_eq!(&inferences, &expected);
     }
+
+    #[test]
+    fn sum_of_consecutive_ints_is_odd() {
```
Comment (bddap) [inline src/infer.rs:191]: Thanks for adding this.

**Comment by nmrshll** [comment]
Thanks for the explanations and comments. 
Looks good to merge now I would say.

**Comment by bddap** [comment]
Tagged and pushed changes to https://www.npmjs.com/package/rify and https://crates.io/crates/rify

**Pull request merged**